stdlib/asm/math/fri.masm

```diff
@@ -0,0 +1,233 @@
+# will be at locaddr.0. The commitments will be at the subsequent addresses. The total 
+# number of such commitments is t_d - 3 (excluding the remainder).
+proc.preprocess
+    adv_push.4                  #[num_q,d,g,t_d,add,..]
+    dup.3                       #[t_d,num_q,d,g,t_d,add,..]
+    movdn.4                     #[num_q,d,g,t_d,t_d,add,..]
+    dup.5                       #[add,num_q,d,g,t_d,t_d,add,..]
+    mem_storew dropw            #[t_d,add,..]
+    dup
+    u32checked_neq.3            #[?,t_d-1,add,..]
+    while.true
+        push.0.0.0.0
+        adv_loadw                       #[C,t_d,add,..]
+        movup.5                         #[add,C,t_d,..]
+        u32wrapping_add.1               #[add+1,C,t_d,..]
+        dup movdn.6                     #[add+1,C,t_d,add+1,..]
+        mem_storew                      #[C,t_d,add+1,..]
+        adv_loadw                       #[0,0,a1,a0,t_d,add+1,..]
+        movup.5                         #[add+1,0,0,a1,a0,t_d,..]
+        u32wrapping_add.1               #[add+2,0,0,a1,a0,t_d,..]
+        dup movdn.6                     #[add+2,0,0,a1,a0,t_d,add+2,..]
+        mem_storew dropw                #[t_d,add+2,..]
+        u32wrapping_sub.1 dup           #[t_d-1,t_d-1,add+2,..]
+        u32checked_neq.3                #[?,t_d-1,add+2,..]
+    end
+    #[0,add,..]
+    drop
+    # value of the remainder codeword
+    adv_push.2
+    movup.2
+end
+
+# Input: [t_d,e1,e0,p,d,poe,add',..]
+# Output: [d,p,C,t_d,e1,e0,poe,a1,a0,..]
+export.prepare_next_new
+    push.0.0.0.0                           #[0,0,0,0,t_d,e1,e0,p,d,poe,add',..]
+    movup.10 sub.1 dup                     #[add'-1,add'-1,0,0,0,0,t_d,e1,e0,p,d,poe,..]
+    movdn.11                               #[add'-1,0,0,0,0,t_d,e1,e0,p,d,poe,add'-1,..]
+    mem_loadw drop drop                    #[a1,a0,t_d,e1,e0,p,d,poe,add'-1,..]
+    swapw                                  #[e0,p,d,poe,a,a,t_d,e1,add'-1,..]
+    push.0.0.0.0 movup.12                  #[add'-1,0,0,0,0,e0,p,d,poe,a,a,t_d,e1,..]
+    sub.1 dup                              #[add'-2,add'-2,0,0,0,0,e0,p,d,poe,a,a,t_d,e1,..]
+    movdn.13 mem_loadw                     #[C,e0,p,d,poe,a,a,t_d,e1,add'-2,..]
+    movup.5 movup.6                        #[d,p,C,e0,poe,a,a,t_d,e1,add'-2,..]
+    movup.11                               #[e1,d,p,C,e0,poe,a,a,t_d,add'-2,..]
+    movdn.6                                #[d,p,C,e1,e0,poe,a,a,t_d,add'-2,..]
+    movup.11                               #[t_d,d,p,C,e1,e0,poe,a,a,add'-2,..]
+    movdn.6                                #[d,p,C,t_d,e1,e0,poe,a1,a0,add'-2,..]
+end
+
+# Given a stack in the following initial configuration [d,p,C,t_d,e1,e0,poe,a1,a0,..] the following
+# procedure computes an iteration of fri verification for a query.
+# TODO: Check where/if some checks are needed in the beginning, like p < d (p&d are u32) and that 2^t_d == d 
+export.verify_query_layer
```

Comment:
Does this verify a single query or one layer of a single query? It seems like the latter. If so, I'd prefer to have another procedure for verifying a single query. Basically, we could have the following call hierarchy:
```
verify -> verify_query -> verify_query_layer
```
In the above:
- `verify` would verify the whole FRI proof.
- `verify_query` would verify a single query.
- `verify_query_layer` would verify a single layer for a query (which I think what the current procedure does).

And at the end, we could also have `verify_remainder`.

Reply:
Makes sense, thank you!